File: week2/14501/14501.cpp
#include <iostream>
#include <vector>

using namespace std;

int N = 0;
int ans = 0;
vector<pair<int, int> >schedule;
void pick_schedule(int, int);

int main() {

    cin >> N;
    schedule.reserve(N);
    int t, p;
    for (int i=0; i<N; i++) {
        cin >> t >> p;
        schedule.push_back(make_pair(t, p));
    }

    pick_schedule(0, 0);
    cout << ans << endl;
}

void pick_schedule(int idx, int sum) {
    ans = max(ans, sum);
    
    if (idx >= N)
        return;

    int t = schedule[idx].first;
    int p = schedule[idx].second;

    if (idx+t <= N)
        pick_schedule(idx+t, sum+p);
    if (idx+1 <= N)
        pick_schedule(idx+1, sum);
}

File: week2/14501/14501.py
class P_14501:
    def __init__(self):
        self.N = int(input())
        self.schedule = [0 for _ in range(self.N)]
        for i in range(self.N):
            self.schedule[i] = tuple(map(int, input().split()))
        self.ans = 0
    
    def run(self):
        self.pick_schedule(0, 0)
        print(self.ans)

    def pick_schedule(self, idx, sum):
        self.ans = max(self.ans, sum)

        if idx >= self.N:
            return

        t = self.schedule[idx][0]
        p = self.schedule[idx][1]

        if idx+t <= self.N:
            self.pick_schedule(idx+t, sum+p) # pick
        if idx+1 <= self.N:
            self.pick_schedule(idx+1, sum) # unpick


if __name__ == "__main__":
    P = P_14501()
    P.run()